mgr/dashboard: carbon tree component

ivoalmeida opened the pull request:
Replaces the deprecated npm package '@circlon/angular-tree-component' by Carbon Tree component.

Fixes: https://tracker.ceph.com/issues/68249





<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [x] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

![image](https://github.com/user-attachments/assets/2fcbb2d8-6515-4592-af65-05aa463781b1)
![image](https://github.com/user-attachments/assets/5a255b35-0417-4365-819a-dcca38f39dbc)
![image](https://github.com/user-attachments/assets/b58cf923-0ba0-4780-ad7e-90016236f094)

ivoalmeida commented on src/pybind/mgr/dashboard/frontend/package.json:
```diff
@@ -64,7 +63,7 @@
     "async-mutex": "0.2.4",
     "bootstrap": "5.2.3",
-    "carbon-components-angular": "5.25.1",
+    "carbon-components-angular": "5.45.1",
```
PTAL https://github.com/ceph/ceph/pull/60626

nizamial09 commented:
Looks like the unit tests are failing: https://jenkins.ceph.com/job/ceph-pull-requests/146280/

ivoalmeida commented:
> Looks like the unit tests are failing: https://jenkins.ceph.com/job/ceph-pull-requests/146280/

Hi @nizamial09 

Tests are passing now.

nizamial09 commented:
Thank you @ivoalmeida. Looks great and sorry for the delay in testing this. Pointing out some things i saw while testing.

1. CephFS directories

If I select a directory and create a snapshot on it, I am not seeing the value there in the table immediately. I have to go to another tree entry and come back to see the table gets updated. It happens with the quota as well. It could be an unrelated issue but I remember it used to work when I was converting the forms to carbon.

https://github.com/user-attachments/assets/938b4042-a74b-4b1c-a241-863653d6b2bf

2. Multi-site topology viewer
I can't seem to do the actions there. Edit opens up the modal but its broken. I can click the Delete action but it doesn't do anything. I saw the below in console log

```
Uncaught TypeError: Cannot read properties of undefined (reading 'data')
    at RgwMultisiteZoneFormComponent.ngOnInit (rgw-multisite-zone-form.component.ts:184:36)
    at callHook (core.mjs:2434:22)
    at callHooks (core.mjs:2403:17)
    at executeInitAndCheckHooks (core.mjs:2354:9)
    at refreshView (core.mjs:10341:21)
```


https://github.com/user-attachments/assets/29291a67-5bf7-47b5-90d6-1c6d03bf8b69


Incase you want to see more entries in the multisite view, you can [simply do this](https://github.com/rhcs-dashboard/ceph-dev?tab=readme-ov-file#rgw-multi-site). Just setting the `RGW_MULTISITE=1` in .env file and starting the cluster normally `dc up -d ceph` should be fine for the basic testing of the topology viewer. I don't think you'll need to start 2 clusters there.

ivoalmeida commented:
> Thank you @ivoalmeida. Looks great and sorry for the delay in testing this. Pointing out some things i saw while testing.
> 
> 1. CephFS directories
> 
> If I select a directory and create a snapshot on it, I am not seeing the value there in the table immediately. I have to go to another tree entry and come back to see the table gets updated. It happens with the quota as well. It could be an unrelated issue but I remember it used to work when I was converting the forms to carbon.
> 
>  blob.5.mp4 
> 2. Multi-site topology viewer
>    I can't seem to do the actions there. Edit opens up the modal but its broken. I can click the Delete action but it doesn't do anything. I saw the below in console log
> 
> ```
> Uncaught TypeError: Cannot read properties of undefined (reading 'data')
>     at RgwMultisiteZoneFormComponent.ngOnInit (rgw-multisite-zone-form.component.ts:184:36)
>     at callHook (core.mjs:2434:22)
>     at callHooks (core.mjs:2403:17)
>     at executeInitAndCheckHooks (core.mjs:2354:9)
>     at refreshView (core.mjs:10341:21)
> ```
> 
>  blob.4.mp4 
> Incase you want to see more entries in the multisite view, you can [simply do this](https://github.com/rhcs-dashboard/ceph-dev?tab=readme-ov-file#rgw-multi-site). Just setting the `RGW_MULTISITE=1` in .env file and starting the cluster normally `dc up -d ceph` should be fine for the basic testing of the topology viewer. I don't think you'll need to start 2 clusters there.

Thanks @nizamial09 
I think I have addressed the issues you spotted. PTAL

nizamial09 commented:
tested again, thanks for addressing the other comments.

I found 2 issues again but could be minor ones. Apart from that it looks good to me.

- when we have many items in the tree than it should in the view, shouldn't it have a scroll bar?
![image](https://github.com/user-attachments/assets/adf1bdb3-0196-45e0-b8e0-c6222641b8f2)

- I created a zonegroup on a realm, and whenever i refresh the topology viewer page (sometimes i just have to go back to other pages and come back), and it shows this ExpressionChanged error in console

![image](https://github.com/user-attachments/assets/94ce8a68-55a2-4b0d-870a-349d25d1326d)

ivoalmeida commented:
> I can see the scroll bar in directories page and also the zonegroup creation issue is gone.
> 
> But I don't see scrollbar for multi-site view and I feel like the multisite topology viewer is stuttery when I scroll the page down and up. Also I seen the page not loading anything when I move away from the page and comeback. I need to refresh the page.
> 
>  blob.7.mp4 
> And I saw that iSCSI topology viewer doesn't show the item name. ![image](https://private-user-images.githubusercontent.com/71764184/389490537-163de7ce-2695-4809-a595-b4da52cb3c6b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI1NDAzOTgsIm5iZiI6MTczMjU0MDA5OCwicGF0aCI6Ii83MTc2NDE4NC8zODk0OTA1MzctMTYzZGU3Y2UtMjY5NS00ODA5LWE1OTUtYjRkYTUyY2IzYzZiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTI1VDEzMDgxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU1NzkwYTEzNDMwMjdiMTFkZGEyMDNmNDc0ZmRiMWZlNDU3ODIwN2UyMmViNWNiNjVhMmMxZDdkYjg0ODI2NjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.yLhy8WrwvMxdG2RtQbDuXwE8PAB8n81H-5HS52LqxsQ)

Hi @nizamial09 thanks for spotting these issues. The old tree component did a lot of heavy lifting to optimise performance. I did see the CPU utilisation was really high and I fixed that by changing the change detection strategy to OnPush and detect changes only when new data is updated. Now CPU utilisation is very low which fantastic. the other issue was the data structure of context had to be put inside a data property instead of spreading them all in the root of each node element. This was fixed.

nizamial09 commented:
@ivoalmeida thank you for the explanation and the fix. Its a lot better now :partying_face: Although, scroll bar is still not there in multi-site topology viewer and the name is not there in iSCSI topology viewer either.

nizamial09 commented:
jenkins test make check

nizamial09 commented:
jenkins test make check arm64

nizamial09 commented:
jenkins test api

nizamial09 reviewed (approved):
works perfectly now, thank you :partying_face:

ivoalmeida commented:
> @ivoalmeida unit tests [are failing btw](https://jenkins.ceph.com/job/ceph-pull-requests/147463/testReport/junit/projectroot.src.test/mgr/mgr_dashboard_frontend_unittests/)

They should be passing now. Thanks

ivoalmeida commented:
jenkins test api
